salt/top.sls.in

```diff
@@ -1,22 +1,51 @@
-metalk8s-@@VERSION:
-  'metalk8s:node:version:@@VERSION':
-    - match: pillar
-    - metalk8s.roles.minion
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:bootstrap':
-    - match: compound
-    - metalk8s.roles.bootstrap
-    - metalk8s.roles.salt-master
-    - metalk8s.roles.registry
-    - metalk8s.roles.repository
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:ca':
-    - match: compound
-    - metalk8s.roles.ca
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:etcd':
-    - match: compound
-    - metalk8s.roles.etcd
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:master':
-    - match: compound
-    - metalk8s.roles.master
-  'I@metalk8s:node:version:@@VERSION and I@metalk8s:node:roles:node':
-    - match: compound
-    - metalk8s.roles.node
+{%- set version = "@@VERSION" %}
+
+{%- set version_match = "metalk8s:nodes:" ~ grains['id'] ~ ":version:" ~ version %}
+
+{% macro role_match(name) -%}
+metalk8s:nodes:{{ grains['id'] }}:roles:{{ name }}
+{%- endmacro %}
+
+{% macro role(name='', states=[]) -%}
+{%- if name %}
+  'I@{{ version_match }} and I@{{ role_match(name) }}':
+{%- else %}
+  'I@{{ version_match }}':
+{%- endif %}
+    {{ ([{ "match": "compound" }] + states) | yaml }}
+{%- endmacro %}
```

Comment:
Meh, dunno. I did that at first, but then you make assumptions about which *test* the macro renders (the `match: compound` is still required). That felt 'wrong', and actually caused some 'bugs' in the beginning.

Basically, if the `role` macro could be rendered in a single line, yes. However, it needs *2* lines: the 'check' line, and then `- match: compound` beneath it...

Reply:
Yes that need both line but it's ok I think, what kind of bug ?

Reply:
I think it was basically that, in your suggested syntax, one needs to write 'invalid'-looking YAML:

```
[...]
  {{ role('bootstrap') }}
    - foo
    - bar
```

When writing YAML, one would expect:

```
  {{ role('bootstrap') }}:
    - foo
    - bar
```

However that `:` would be invalid. And we can't "not render the `:` in the macro", because we need the second line.

Reply:
Yes I agree it's weird syntax at the end but btw our current macro like `pkg_installed` it's the same so ... I don't